File: .agents/skills/lisp-brackets/check-brackets.py
#!/usr/bin/env python3

"""Lisp bracket checker for .meta files.

Independent of any project build tools. Validates that every file
has balanced brackets, ignoring content inside strings and
semicolon comments.

Usage:
    python3 check-brackets.py <file1.meta> [file2.meta ...]
    python3 check-brackets.py src/**/*.meta

Exit code: 1 if any file has unbalanced brackets.
"""

import os
import sys
from pathlib import Path


def strip_strings_and_comments(line: str) -> str:
    """Remove string contents and semicolon comments from a line,
    returning a mask where only structural brackets remain.

    We can't simply strip because that would shift column positions.
    Instead we replace non-structural chars with spaces so column
    numbers stay accurate.
    """
    result = []
    i = 0
    in_string = False
    escape = False
    prev = '\0'

    while i < len(line):
        c = line[i]

        if escape:
            escape = False
            if in_string:
                result.append(' ')
            else:
                result.append(c)
            i += 1
            prev = c
            continue

        if c == '\\' and in_string:
            escape = True
            result.append(' ')
            i += 1
            prev = c
            continue

        if c == '"':
            in_string = not in_string
            result.append(' ')  # the quote itself is structural but we mask it
            i += 1
            prev = c
            continue

        if in_string:
            result.append(' ')
        elif c == ';':
            # rest of line is comment
            result.extend([' '] * (len(line) - i))
            break
        elif c in '()':
            result.append(c)
        else:
            result.append(' ')

        prev = c
        i += 1

    return ''.join(result)


def check_file(filepath: str) -> bool:
    """Check a single file. Returns True if balanced."""
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except FileNotFoundError:
        print(f'{filepath}: FILE NOT FOUND', file=sys.stderr)
        return True  # not a bracket error
    except Exception as e:
        print(f'{filepath}: READ ERROR — {e}', file=sys.stderr)
        return True

    depth = 0
    last_open_line = None
    last_open_col = None

    for lineno, raw_line in enumerate(lines, start=1):
        masked = strip_strings_and_comments(raw_line)

        for col, c in enumerate(masked):
            if c == '(':
                depth += 1
                last_open_line = lineno
                last_open_col = col
            elif c == ')':
                depth -= 1
                if depth < 0:
                    # Negative depth: extra closing bracket
                    ruler = ' ' * col + '^'
                    print(f'{filepath}:{lineno}:{col + 1}  EXTRA ) — depth went negative (-1)')
                    print(f'  L{lineno}: {raw_line.rstrip()}')
                    print(f'         {ruler}')
                    return False

    if depth > 0:
        # Positive residual: unmatched open brackets
        print(f'{filepath}:{len(lines)}  UNMATCHED OPEN — {depth} open bracket(s) never closed')
        if last_open_line is not None:
            print(f'  Last opened: L{last_open_line}:{last_open_col + 1 if last_open_col is not None else "?"}')
            if last_open_line <= len(lines):
                print(f'  L{last_open_line}: {lines[last_open_line - 1].rstrip()}')
        return False

    open_count = sum(1 for line in lines for c in strip_strings_and_comments(line) if c == '(')
    close_count = sum(1 for line in lines for c in strip_strings_and_comments(line) if c == ')')
    print(f'{filepath}: OK ({open_count} open, {close_count} close, depth 0)')
    return True


def main() -> int:
    if len(sys.argv) < 2:
        print('Usage: python3 check-brackets.py <file1.meta> [file2.meta ...]', file=sys.stderr)
        print('       python3 check-brackets.py src/**/*.meta', file=sys.stderr)
        return 1

    all_ok = True
    for arg in sys.argv[1:]:
        p = Path(arg)
        # Direct paths (absolute or with wildcards like /dev/stdin)
        if p.exists() or os.path.isabs(arg):
            if not check_file(str(p)):
                all_ok = False
        else:
            paths = sorted(Path('.').glob(arg))
            if not paths:
                print(f'{arg}: NO FILES MATCHED', file=sys.stderr)
                all_ok = False
                continue
            for mp in paths:
                if not check_file(str(mp)):
                    all_ok = False

    return 0 if all_ok else 1


if __name__ == '__main__':
    sys.exit(main())

File: skills/lisp-brackets/check-brackets.mjs
#!/usr/bin/env node

import { readFileSync, existsSync, readdirSync, statSync } from "node:fs"
import { resolve, relative, isAbsolute, dirname, basename, extname, join } from "node:path"
import { fileURLToPath } from "node:url"

const scriptDir = dirname(fileURLToPath(import.meta.url))

function stripLine(line) {
  const result = []
  let i = 0
  let inString = false
  let escape = false

  while (i < line.length) {
    const c = line[i]

    if (escape) {
      escape = false
      result.push(inString ? " " : c)
      i++
      continue
    }

    if (c === "\\" && inString) {
      escape = true
      result.push(" ")
      i++
      continue
    }

    if (c === '"') {
      inString = !inString
      result.push(" ")
      i++
      continue
    }

    if (inString) {
      result.push(" ")
    } else if (c === ";") {
      result.push(...Array(line.length - i).fill(" "))
      break
    } else if (c === "(" || c === ")") {
      result.push(c)
    } else {
      result.push(" ")
    }

    i++
  }

  return result.join("")
}

function checkFile(filepath) {
  let lines
  try {
    lines = readFileSync(filepath, "utf-8").split("\n")
  } catch (e) {
    if (e.code === "ENOENT") {
      process.stderr.write(`${filepath}: FILE NOT FOUND\n`)
      return true
    }
    process.stderr.write(`${filepath}: READ ERROR — ${e.message}\n`)
    return true
  }

  let depth = 0
  let lastOpenLine = null
  let lastOpenCol = null

  for (let lineno = 0; lineno < lines.length; lineno++) {
    const rawLine = lines[lineno]
    const masked = stripLine(rawLine)

    for (let col = 0; col < masked.length; col++) {
      const c = masked[col]
      if (c === "(") {
        depth++
        lastOpenLine = lineno + 1
        lastOpenCol = col
      } else if (c === ")") {
        depth--
        if (depth < 0) {
          const ruler = " ".repeat(col) + "^"
          console.log(`${filepath}:${lineno + 1}:${col + 1}  EXTRA ) — depth went negative (-1)`)
          console.log(`  L${lineno + 1}: ${rawLine}`)
          console.log(`         ${ruler}`)
          return false
        }
      }
    }
  }

  if (depth > 0) {
    console.log(`${filepath}:${lines.length}  UNMATCHED OPEN — ${depth} open bracket(s) never closed`)
    if (lastOpenLine !== null) {
      const colStr = lastOpenCol !== null ? `${lastOpenCol + 1}` : "?"
      console.log(`  Last opened: L${lastOpenLine}:${colStr}`)
      if (lastOpenLine <= lines.length) {
        console.log(`  L${lastOpenLine}: ${lines[lastOpenLine - 1]}`)
      }
    }
    return false
  }

  let openCount = 0
  let closeCount = 0
  for (const line of lines) {
    const masked = stripLine(line)
    for (const c of masked) {
      if (c === "(") openCount++
      if (c === ")") closeCount++
    }
  }
  console.log(`${filepath}: OK (${openCount} open, ${closeCount} close, depth 0)`)
  return true
}

function glob(pattern, baseDir) {
  const results = []
  const parts = pattern.split("/")

  function walk(dir, idx) {
    if (idx >= parts.length) {
      results.push(dir)
      return
    }

    const part = parts[idx]

    if (part === "**") {
      const entries = readdirSync(dir, { withFileTypes: true })
      for (const entry of entries) {
        const full = join(dir, entry.name)
        if (entry.isDirectory()) {
          walk(full, idx)
          walk(full, idx + 1)
        } else if (idx === parts.length - 1) {
          results.push(full)
        }
      }
      return
    }

    if (part.includes("*")) {
      const regex = new RegExp(
        "^" + part.replace(/\./g, "\\.").replace(/\*/g, ".*") + "$"
      )
      const entries = readdirSync(dir, { withFileTypes: true })
      const isLast = idx === parts.length - 1
      for (const entry of entries) {
        if (regex.test(entry.name)) {
          if (isLast && entry.isFile()) {
            results.push(join(dir, entry.name))
          } else if (!isLast && entry.isDirectory()) {
            walk(join(dir, entry.name), idx + 1)
          }
        }
      }
      return
    }

    const next = join(dir, part)
    if (existsSync(next)) {
      walk(next, idx + 1)
    }
  }

  walk(baseDir, 0)
  return results
}

function main() {
  if (process.argv.length < 3) {
    process.stderr.write("Usage: node check-brackets.mjs <file1.meta> [file2.meta ...]\n")
    process.stderr.write("       node check-brackets.mjs src/**/*.meta\n")
    process.exit(1)
  }

  let allOk = true

  for (const arg of process.argv.slice(2)) {
    if (isAbsolute(arg) || existsSync(arg)) {
      if (!checkFile(arg)) allOk = false
    } else if (arg.includes("*")) {
      const paths = glob(arg, ".")
      if (paths.length === 0) {
        process.stderr.write(`${arg}: NO FILES MATCHED\n`)
        allOk = false
        continue
      }
      for (const p of paths.sort()) {
        if (!checkFile(p)) allOk = false
      }
    } else {
      if (existsSync(arg)) {
        if (!checkFile(arg)) allOk = false
      } else {
        process.stderr.write(`${arg}: FILE NOT FOUND\n`)
        allOk = false
      }
    }
  }

  process.exit(allOk ? 0 : 1)
}

main()
